### ISSUE TITLE
How can I define global variables that are accessible across different requests?

### ISSUE THREAD

**Issue by xiaodiannao** [tiangolo/fastapi#592]
Hi everyone, I'm still experimenting with FastAPI and I have a question regarding global variables:

How can I define global variables that are accessible across different requests?

Is it possible to do this?

I tried to add a simple counter `counter = 0` in the main.py and then incrementing it by one in one request and reading back the value in another separate request. The second request returns 0 instead of 1 however. I hope somebody can help me out :)

**Comment by dmontagu** [comment]
@xiaodiannao This should work now (though it won't work if you have separate processes, but that wouldn't work with any kind of global variable). Could you share a code example where this isn't working for you?

**Comment by HacKanCuBa** [comment]
Another solution that might be better than using global vars is to append the information you need to share as a property of `app` (being `app` a `FastAPI()` object).

I.E.:

```python
app = FastAPI()
app.counter = 0

# ....

app.counter += 1
```

**Comment by xiaodiannao** [comment]
Hi @dmontagu and @HacKanCuBa thank you both for your replies! I tried your solution @HacKanCuBa and I'm setting `app.counter = 0` and increment it using `main.app.counter += 1` for each request. In another endpoint (using a separate request) I'm reading out the value of the counter and it is still 0. It's very weird...

**Comment by dmontagu** [comment]
@xiaodiannao 
```python
from fastapi import FastAPI
from starlette.testclient import TestClient

app = FastAPI()

app.counter = 0


@app.get("/increment")
def increment():
    app.counter += 1


@app.get("/status")
def read_status():
    return app.counter


client = TestClient(app)

print(client.get("/status").content)
client.get("/increment")
print(client.get("/status").content)
client.get("/increment")
print(client.get("/status").content)
"""
b'0'
b'1'
b'2'
"""
```
seems to work fine, so I'm guessing this is a bug in your code?

If you can share more we might be able to help you.

**Comment by dmontagu** [comment]
You'll only have multiple processes if you are using a server that runs multiple processes (e.g., gunicorn/uvicorn set up for production).

Typically in development you'd only run a single server process.

**Comment by xiaodiannao** [comment]
Ah I found out what the problem is, my code is more or less the same as yours but I have FastAPI running behind nginx. When I make the requests directly to FastAPI (bypassing nginx) the counter is incremented on the status request. But I don't quite get why this makes a difference (accessing directly vs. over nginx)

**Comment by HacKanCuBa** [comment]
> Ah I found out what the problem is, my code is more or less the same as yours but I have FastAPI running behind nginx. When I make the requests directly to FastAPI (bypassing nginx) the counter is incremented on the status request. But I don't quite get why this makes a difference (accessing directly vs. over nginx)
> 
> Oh and also I'm using the official Docker image tiangolo/uvicorn-gunicorn-fastapi to run my experiments, maybe the problem is that multiple servers are started there? @dmontagu

You shouldn't  bypass nginx, and use gunicorn+uvicorn workers to run the app, setting at least 2 workers.
I'm not sure what's not working for you...
As a side note, latest FastAPI has a new `state` property which might be helpful: https://github.com/tiangolo/fastapi/releases/tag/0.41.0 (see https://github.com/tiangolo/fastapi/pull/502/files for details)

**Comment by xiaodiannao** [comment]
Hey @HacKanCuBa do you know how I can adjust the number of workers? Right now I have 4 workers which is the number of cores of my machine. Could this be the problem, that all of those 4 workers have their own instance of the global counter variable?

**Comment by dmontagu** [comment]
@xiaodiannao Yes this could be the problem. If you make enough requests though (more than the number of workers -- probably 4) you should see the counters increase.

But this design is inherently non-scalable -- you can't have more than a single node if you want the counter to behave properly.

A more robust design would be to store the counter using a cache like redis. If you don't want to deal with configuring more dependencies, you could just use a database table as a key-value store, and store the counter there (and update it in the database on each request). That would scale to any number of processes / nodes / etc.

**Comment by xiaodiannao** [comment]
@dmontagu ah yes that makes sense. I'll try a different approach next, this was just a small test project to try out FastAPI. Thanks a lot for your help!

**Comment by HacKanCuBa** [comment]
Exactly, because that *is not* concurrent nor thread safe. Use Redis' INCR/DECR which it is and *have less headaches* ™  
I can point you, as sample code, something I wrote w/ some friends: https://git.rlab.be/sysadmins/yog_sothoth/blob/master/backend/yog_sothoth/objects/rate_limit.py#L36

Cheers!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue @xiaodiannao  :+1:
